src/types.ts

```diff
@@ -0,0 +1,28 @@
+export interface WebSocket {
+  protocol: string;
+  id?: string;
+
+  close(code: number, reason: string): Promise<void> | void;
+
+  on(
+    message: string,
+    listener: (data: any, ack?: (data?: any) => Promise<void> | void) => void,
+  ): void;
+
+  emit(message: string, data: any): Promise<void> | void | boolean;
+  // onMessage(cb: (data: string) => Promise<void>): void;
+}
+
+export type MessageType =
+  | 'authenticate'
+  | 'subscribe'
+  | 'unsubscribe'
+  | 'connect'
+  | 'disconnect'
+  | `ack:${number}`;
```

Comment:
what are acks used/needed for?

Reply:
testing mostly

Reply:
shouldn't be `ack` though, and then `number` be part of the payload? because otherwise you will need to listen to multiple topics

Reply:
it's easier to do the state tracking, when you emit an event you immediately just listen `.once` for the specific ack event. It's functionally the similiar as having a single handler since this is client<->server they are all listening to a single "message" event on the socket any way.

Reply:
yes, with once makes sense